cartoframes/dataset.py

```
@@ -55,52 +58,61 @@ def __init__(self, table_name=None, schema='public', query=None, df=None, gdf=No
+        if table_name:
+            self.table_name = normalize_name(table_name)
+        if schema:
+            self.schema = schema
+        if context:
+            self.cc = context
+
+        if self.table_name is None or self.cc is None:
+            raise CartoException('You should provide a table_name and context to upload data.')
+
+        if self.gdf is None and self.df is None and self.query is None:
+            raise CartoException('Nothing to upload.')
 
-            if not self.exists():
+        already_exists_error = NameError('Table with name {table_name} already exists in CARTO.'
+                                         ' Please choose a different `table_name` or use'
+                                         ' if_exists="replace" to overwrite it'.format(table_name=self.table_name))
+
+        # priority order: gdf, df, query
+        if self.gdf is not None:
+            warn('GeoDataFrame option is still under development. We will try the upload with DataFrame')
+
+        if self.df is not None:
+            if not self.exists() or if_exists == Dataset.REPLACE:
                 self._create_table(with_lonlat)
-            else:
-                if if_exists == Dataset.FAIL:
-                    raise NameError(('Table with name {table_name} already exists in CARTO.'
-                                     ' Please choose a different `table_name` or use'
-                                     ' if_exists="replace" to overwrite it').format(table_name=self.table_name))
-                elif if_exists == Dataset.REPLACE:
-                    self._create_table(with_lonlat)
+            elif if_exists == Dataset.FAIL:
+                raise already_exists_error
+
             self._copyfrom(with_lonlat)
 
+        elif self.query is not None:
+            if not self.exists() or if_exists == Dataset.REPLACE:
+                self._create_table_from_query()
+            elif if_exists == Dataset.FAIL:
+                raise already_exists_error
+
```

Since we do support `Dataset.APPEND` for (geo)dataframes, are we considering the `Dataset.APPEND` scenario for `query`? (Not saying to do it in this PR, but let's extract a new issue at least)

---

I am not sure if it makes sense (at least I don't see the use case right now)

---

If we don't support it, we should raise an error for that combination of parameters.

---

Done